File: src/lexer.c
#include "lexer.h"

#include "allocator.h"

#include <assert.h>
#include <ctype.h>
#include <inttypes.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>

struct string const token_names[]
    = { [TOKEN_NONE]         = {
        .data = "TOKEN_NONE",
        .length = 10,
    },
    [TOKEN_NAME]         = {
        .data = "TOKEN_NAME",
        .length = 10,
    },
    [TOKEN_ASSIGN]         = {
        .data = "TOKEN_ASSIGN",
        .length = 12,
    },
    [TOKEN_DOT]         = {
        .data = "TOKEN_DOT",
        .length = 9,
    },
    [TOKEN_DOT_DOT_DOT]         = {
        .data = "TOKEN_DOT_DOT_DOT",
        .length = 17,
    },
    [TOKEN_COMMA]         = {
        .data = "TOKEN_COMMA",
        .length = 11,
    },
    [TOKEN_CHAR]         = {
        .data = "TOKEN_CHAR",
        .length = 10,
    },
    [TOKEN_STRING]         = {
        .data = "TOKEN_STRING",
        .length = 12,
    },
    [TOKEN_OPEN_BRACE]         = {
        .data = "TOKEN_OPEN_BRACE",
        .length = 16,
    },
    [TOKEN_CLOSE_BRACE]         = {
        .data = "TOKEN_CLOSE_BRACE",
        .length = 17,
    },
    [TOKEN_OPEN_PAREN]         = {
        .data = "TOKEN_OPEN_PAREN",
        .length = 16,
    },
    [TOKEN_CLOSE_PAREN]         = {
        .data = "TOKEN_CLOSE_PAREN",
        .length = 17,
    },
    [TOKEN_GREATER_THAN]         = {
        .data = "TOKEN_GREATER_THAN",
        .length = 18,
    },
    [TOKEN_SEMICOLON]         = {
        .data = "TOKEN_SEMICOLON",
        .length = 15,
    },
    [TOKEN_INTEGER]         = {
        .data = "TOKEN_INTEGER",
        .length = 13,
    },
    [TOKEN_EOF]         = {
        .data = "TOKEN_EOF",
        .length = 9,
    }
};

static noreturn void
error(struct lexer lexer[static const 1], char const* const format, ...) {
    fprintf(
        stderr, "%s:%d:%d: syntax error: ", "vix", lexer->location.line_number,
        lexer->location.column_number
    );

    va_list ap;
    va_start(ap, format);
    vfprintf(stderr, format, ap);
    va_end(ap);

    fprintf(stderr, "\n");
    error_line(lexer->allocator, lexer->location);
    exit(EXIT_LEX);
}

struct lexer
lexer_new(
    struct allocator allocator[static const 1], FILE f[static const 1],
    i32 file_id
) {
    return (struct lexer){
        .allocator   = allocator,
        .in          = f,
        .buffer_size = 256,
        .buffer      = allocator_allocate(allocator, 256),
        .un =
            (struct token){
                             .type = TOKEN_NONE,
                             },
        .location =
            (struct location){
                             .line_number   = 1,
                             .column_number = 0,
                             .file          = file_id,
                             },
        .c =
            {
                             (struct codepoint){.ok = false},
                             (struct codepoint){.ok = false},
                             },
    };
}

static void
lexer_clear_buffer(struct lexer lexer[static const 1]) {
    lexer->buffer_length = 0;
    lexer->buffer[0]     = 0;
}

void
lexer_finish(struct lexer lexer[static const 1]) {
    fclose(lexer->in);
}

static void
append_buffer(
    struct lexer lexer[static const 1], struct utf8char const utf8_char
) {
    if (lexer->buffer_length + utf8_char.length >= lexer->buffer_size) {
        usize const old_buffer_size = lexer->buffer_size;
        do {
            lexer->buffer_size *= 2;
        } while (lexer->buffer_length + utf8_char.length >= lexer->buffer_size);
        lexer->buffer = allocator_resize(
            lexer->allocator, lexer->buffer, old_buffer_size, lexer->buffer_size
        );
    }
    memcpy(
        lexer->buffer + lexer->buffer_length, &utf8_char.characters,
        utf8_char.length
    );
    lexer->buffer_length += utf8_char.length;
    lexer->buffer[lexer->buffer_length] = '\0';
}

static void
update_line_number(
    struct location location[static const 1], struct codepoint const cp
) {
    if (!cp.ok) {
        return;
    }
    if (cp.character == '\n') {
        location->line_number += 1;
        location->column_number = 0;
    } else if (cp.character == '\t') {
        location->column_number += 4;
    } else {
        location->column_number += 1;
    }
}

static struct codepoint
next(
    struct lexer lexer[static const 1], struct location location[const 1],
    bool const is_buffer
) {
    struct codepoint character;
    if (lexer->c[0].ok) {
        character      = lexer->c[0];
        lexer->c[0]    = lexer->c[1];
        lexer->c[1].ok = false;
    } else {
        character = utf8_get(lexer->in);
        update_line_number(&lexer->location, character);
        if (!character.ok && !feof(lexer->in)) {
            error(lexer, "Invalid UTF-8 character");
        }
    }

    if (location != nullptr) {
        *location = lexer->location;
        for (i32 i = 0; i < 2 && lexer->c[i].ok; ++i) {
            update_line_number(&lexer->location, lexer->c[i]);
        }
    }

    if (!character.ok || !is_buffer) {
        return character;
    }
    struct utf8char const encode_result = utf8_encode(character);
    append_buffer(lexer, encode_result);
    return character;
}

static bool
is_whitespace(u8 c) {
    switch (c) {
        case '\t':
        case '\n':
        case '\r':
        case ' ':
            return true;
    }
    return false;
}

static struct codepoint
wgetc(
    struct lexer lexer[static const 1], struct location location[static const 1]
) {
    struct codepoint cp;
    do {
        cp = next(lexer, location, false);
    } while (cp.ok && is_whitespace(cp.character));
    return cp;
}

static void
consume(struct lexer lexer[static const 1], usize const n) {
    for (usize i = 0; i < n; i += 1) {
        while ((lexer->buffer[--lexer->buffer_length] & 0xC0) == 0x80)
            ;
    }
    lexer->buffer[lexer->buffer_length] = 0;
}

static void
push(
    struct lexer lexer[static const 1], u32 const character,
    bool const is_buffer
) {
    assert(!lexer->c[1].ok);
    lexer->c[1]           = lexer->c[0];
    lexer->c[0].character = character;
    lexer->c[0].ok        = true;
    if (is_buffer) {
        consume(lexer, 1);
    }
}

static void
lex_number(
    struct lexer lexer[static const 1], struct token out[static const 1]
) {
    struct codepoint cp = next(lexer, &out->location, true);
    u32 character       = cp.character;
    assert(cp.ok && character <= 0x7F && isdigit(character));

    static char const* chars = "0123456789";

    do {
        character = cp.character;
        if (strchr(chars, character)) {
            continue;
        } else {
            break;
        }
    } while ((cp = next(lexer, nullptr, true)).ok);
    push(lexer, character, true);

    out->integer = strtoumax((char*) lexer->buffer, nullptr, 10);

    out->type = TOKEN_INTEGER;
    lexer_clear_buffer(lexer);
}

static void
lex_name(struct lexer lexer[static const 1], struct token out[static const 1]) {
    struct codepoint cp = next(lexer, &out->location, true);
    u32 character       = cp.character;
    assert(cp.ok && character <= 0x7F && isalpha(character));
    while (cp.ok) {
        cp        = next(lexer, &out->location, true);
        character = cp.character;
        if (character > 0x7F || (!isalnum(character) && character != '_')) {
            push(lexer, character, true);
            break;
        }
    }

    out->type = TOKEN_NAME;
    out->name = string_duplicate(
        lexer->allocator,
        (struct string){
            .data   = lexer->buffer,
            .length = lexer->buffer_length,
        }
    );
    lexer_clear_buffer(lexer);
}

static struct utf8char
lex_rune(struct lexer lexer[static const 1]) {
    struct codepoint cp = next(lexer, nullptr, false);
    u32 character       = cp.character;
    assert(cp.ok);

    switch (character) {
        case '\\': {
            character = next(lexer, nullptr, false).character;
            switch (character) {
                case '0':
                    return (struct utf8char){
                        .characters = { '\0' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'a':
                    return (struct utf8char){
                        .characters = { '\0' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'b':
                    return (struct utf8char){
                        .characters = { '\b' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'r':
                    return (struct utf8char){
                        .characters = { '\r' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 't':
                    return (struct utf8char){
                        .characters = { '\t' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'n':
                    return (struct utf8char){
                        .characters = { '\n' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'f':
                    return (struct utf8char){
                        .characters = { '\f' },
                        .length     = 1,
                        .ok         = true,
                    };
                case 'v':
                    return (struct utf8char){
                        .characters = { '\v' },
                        .length     = 1,
                        .ok         = true,
                    };
                case '\\':
                    return (struct utf8char){
                        .characters = { '\\' },
                        .length     = 1,
                        .ok         = true,
                    };
                case '\'':
                    return (struct utf8char){
                        .characters = { '\'' },
                        .length     = 1,
                        .ok         = true,
                    };
                case '"':
                    return (struct utf8char){
                        .characters = { '\"' },
                        .length     = 1,
                        .ok         = true,
                    };
                default:
                    error(lexer, "Invalid escape '\\%c'", character);
            }
            vix_unreachable();
        }
        default:
            return utf8_encode(cp); // TODO
    }
}

static void
lex_string(
    struct lexer lexer[static const 1], struct token out[static const 1]
) {
    struct codepoint cp = next(lexer, &out->location, true);
    u32 character       = cp.character;
    u32 delimiter;

    switch (character) {
        case '"':
            delimiter = character;
            cp        = next(lexer, nullptr, false);
            character = cp.character;
            while (character != delimiter) {
                if (!cp.ok) {
                    error(lexer, "Unexpected end of file");
                }
                push(lexer, character, false);
                if (delimiter == '"') {
                    struct utf8char utf8 = lex_rune(lexer);
                    append_buffer(lexer, utf8);
                } else {
                    next(lexer, nullptr, true);
                }
                cp        = next(lexer, nullptr, false);
                character = cp.character;
            }
            out->type   = TOKEN_STRING;
            out->string = string_duplicate(
                lexer->allocator,
                (struct string){
                    .data   = lexer->buffer + 1,
                    .length = lexer->buffer_length,
                }
            );
            lexer_clear_buffer(lexer);
            return;
        default:
            vix_unreachable();
    }
    vix_unreachable();

    lexer_clear_buffer(lexer);
}

enum lex_token_type
lex(struct lexer lexer[static const 1], struct token out[static const 1]) {
    if (lexer->un.type != TOKEN_NONE) {
        *out           = lexer->un;
        lexer->un.type = TOKEN_NONE;
        return out->type;
    }
    struct codepoint cp = wgetc(lexer, &out->location);
    if (!cp.ok) {
        out->type = TOKEN_EOF;
        return out->type;
    }
    u32 character = cp.character;

    if (character <= 0x7F && isdigit(character)) {
        push(lexer, character, false);
        lex_number(lexer, out);
        return TOKEN_INTEGER;
    }

    if (character <= 0x7F && isalpha(character)) {
        push(lexer, character, false);
        lex_name(lexer, out);
        return TOKEN_NAME;
    }

    switch (character) {
        case '"':
            push(lexer, character, false);
            lex_string(lexer, out);
            return TOKEN_STRING;
            break;
        case '.':
            switch ((character = next(lexer, nullptr, false).character)) {
                case '.':
                    switch ((character = next(lexer, nullptr, false).character)
                    ) {
                        case '.':
                            out->type = TOKEN_DOT;
                            break;
                        default:
                            error(lexer, "Unknown sequence '..'");
                    }
                    break;
                default:
                    push(lexer, character, false);
                    out->type = TOKEN_DOT;
                    break;
            }
            break;
        case '=':
            out->type = TOKEN_ASSIGN;
            break;
        case ',':
            out->type = TOKEN_COMMA;
            break;
        case '{':
            out->type = TOKEN_OPEN_BRACE;
            break;
        case '}':
            out->type = TOKEN_CLOSE_BRACE;
            break;
        case '(':
            out->type = TOKEN_OPEN_PAREN;
            break;
        case ')':
            out->type = TOKEN_CLOSE_PAREN;
            break;
        case '>':
            out->type = TOKEN_GREATER_THAN;
            break;
        case ';':
            out->type = TOKEN_SEMICOLON;
            break;
        default:
            error(lexer, "Unexpected character '%d'", character);
    }

    return out->type;
}

void
unlex(struct lexer lexer[static const 1], struct token out[static const 1]) {
    assert(lexer->un.type == TOKEN_NONE);
    lexer->un = *out;
}

void
token_finish(struct token token[static const 1]) {
    token->type     = TOKEN_NONE;
    token->location = (struct location){};
}

File: src/lexer.zig
const std = @import("std");

const util = @import("util.zig");
const Location = util.Location;

pub const LexTokenType = enum {
    none,
    name,
    assign,
    dot,
    dot_dot_dot,
    comma,
    character,
    string,
    open_brace,
    close_brace,
    open_paren,
    close_paren,
    greater_than,
    semicolon,
    integer,
    eof,
};

pub const Token = struct {
    type: LexTokenType,
    location: Location,
    value: union {
        name: []const u8,
        character: u8,
        string: []const u8,
        integer: u64,
    },

    pub fn initEmpty() @This() {
        return .{
            .type = .none,
            .location = std.mem.zeroes(Location),
            .value = .{
                .integer = 0,
            },
        };
    }

    pub fn tokenFinish(self: *@This()) void {
        self.type = .none;
        self.value = .{ .integer = 0 };
        self.location = std.mem.zeroInit(Location, .{});
    }
};

pub const Lexer = struct {
    allocator: std.mem.Allocator,
    source_iterator: std.unicode.Utf8Iterator,
    buffer: std.ArrayList(u8),
    c: [2]?u21,
    un: Token,
    location: Location,

    pub fn init(allocator: std.mem.Allocator, source: []const u8, file_id: u32) !@This() {
        const view = try std.unicode.Utf8View.init(source);
        return .{
            .allocator = allocator,
            .source_iterator = view.iterator(),
            .buffer = std.ArrayList(u8).init(allocator),
            .un = .{
                .location = std.mem.zeroInit(Location, .{}),
                .type = .none,
                .value = .{
                    .integer = 0,
                },
            },
            .location = .{
                .line_number = 1,
                .column_number = 1,
                .file = file_id,
            },
            .c = std.mem.zeroes([2]?u21),
        };
    }

    pub fn unlex(self: *@This(), out: *Token) void {
        std.debug.assert(self.un.type == .none);
        self.un = out.*;
    }

    pub fn lex(self: *@This(), out: *Token) anyerror!LexTokenType {
        if (self.un.type != .none) {
            out.* = self.un;
            self.un.type = .none;
            return out.type;
        }

        var cp = try self.wgetc(&out.location);
        if (cp == null) {
            out.type = .eof;
            return out.type;
        }
        const character = cp.?;
        if (character <= 0x7F and std.ascii.isDigit(@intCast(character))) {
            self.push(character, false);
            try self.lexNumber(out);
            return .integer;
        }
        if (character <= 0x7F and std.ascii.isAlphabetic(@intCast(character))) {
            self.push(character, false);
            try self.lexName(out);
            return .name;
        }

        switch (character) {
            '<' => {
                self.push(character, false);
                try self.lexString(out);
                return .string;
            },
            '.' => {
                cp = try self.next(null, false);
                std.debug.assert(cp != null);
                switch (cp.?) {
                    '.' => {
                        cp = try self.next(null, false);
                        std.debug.assert(cp != null);
                        switch (cp.?) {
                            '.' => out.type = .dot_dot_dot,
                            else => try self.printError("Unknown sequence '..'", .{}),
                        }
                    },
                    else => {
                        self.push(cp.?, false);
                        out.type = .dot;
                    },
                }
            },
            '=' => out.type = .assign,
            ',' => out.type = .comma,
            '{' => out.type = .open_brace,
            '}' => out.type = .close_brace,
            '(' => out.type = .open_paren,
            ')' => out.type = .close_paren,
            '>' => out.type = .greater_than,
            ';' => out.type = .semicolon,
            else => try self.printError("Unexpected character '{}'", .{character}),
        }
        return out.type;
    }

    fn clearBuffer(self: *@This()) void {
        self.buffer.clearRetainingCapacity();
    }

    fn printError(self: *@This(), comptime format: []const u8, args: anytype) anyerror!noreturn {
        const stderr = std.io.getStdErr().writer();
        _ = try stderr.print("{s}:{}:{}: syntax error: ", .{ "vix", self.location.line_number, self.location.column_number });
        _ = try stderr.print(format, args);
        _ = try stderr.print("\n", .{});
        try self.location.errorLine(self.allocator);
        std.process.exit(2);
    }

    fn updateLineNumber(location: *Location, character: ?u21) void {
        if (character == null) {
            return;
        }
        if (character.? == '\n') {
            location.line_number += 1;
            location.column_number = 1;
        } else if (character.? == '\t') {
            location.column_number += 4;
        } else {
            location.line_number += 1;
        }
    }

    fn next(self: *@This(), location: ?*Location, is_buffer: bool) anyerror!?u21 {
        var character: ?u21 = null;
        if (self.c[0] != null) {
            character = self.c[0];
            self.c[0] = self.c[1];
            self.c[1] = null;
        } else {
            character = self.source_iterator.nextCodepoint();
            if (character == null) {
                return null;
                // self.printError("Invalid UTF-8 character", .{});
            }
            updateLineNumber(&self.location, character);
        }

        if (location) |loc| {
            loc.* = self.location;
            var i: usize = 0;
            while (i < 2 and self.c[i] != null) : (i += 1) {
                updateLineNumber(&self.location, self.c[i]);
            }
        }

        if (character == null or !is_buffer) {
            return character;
        }
        var encoded_result = [_]u8{0} ** 4;
        const length = try std.unicode.utf8Encode(character.?, &encoded_result);
        for (encoded_result[0..length]) |char| {
            try self.buffer.append(char);
        }
        return character;
    }

    fn wgetc(self: *@This(), location: ?*Location) !?u21 {
        var cp: ?u21 = null;
        while (true) {
            cp = try self.next(location, false);
            if (cp == null or !std.ascii.isWhitespace(@intCast(cp.?))) {
                break;
            }
        }
        return cp;
    }

    fn consume(self: *@This(), n: usize) void {
        var i: usize = 0;
        while (i < n) : (i += 1) {
            while (self.buffer.items.len != 0) {
                const item = self.buffer.pop();
                if ((item & 0xC0) != 0x80) {
                    break;
                }
            }
        }
    }

    fn push(self: *@This(), character: u21, is_buffer: bool) void {
        std.debug.assert(self.c[1] == null);
        self.c[1] = self.c[0];
        self.c[0] = character;
        if (is_buffer) {
            self.consume(1);
        }
    }

    fn lexNumber(self: *@This(), out: *Token) !void {
        var cp = try self.next(&out.location, true);
        std.debug.assert(cp != null);
        var character = cp.?;
        std.debug.assert(character <= 0x7F and std.ascii.isDigit(@intCast(character)));

        while (true) {
            if (cp == null) {
                break;
            }

            character = cp.?;
            if (!std.ascii.isDigit(@intCast(character))) {
                break;
            }
            cp = try self.next(null, true);
        }
        self.push(character, true);

        out.value = .{
            .integer = try std.fmt.parseInt(u64, self.buffer.items, 10),
        };
        out.type = .integer;
        self.clearBuffer();
    }

    fn lexName(self: *@This(), out: *Token) anyerror!void {
        var cp = try self.next(&out.location, true);
        std.debug.assert(cp != null);
        var character = cp.?;
        std.debug.assert(character <= 0x7F and std.ascii.isAlphabetic(@intCast(character)));
        while (cp != null) {
            cp = try self.next(&out.location, true);
            if (cp) |codepoint| {
                character = codepoint;
                if (character > 0x7F or (!std.ascii.isAlphanumeric(@intCast(character)) and character != '_')) {
                    self.push(character, true);
                    break;
                }
            }
        }

        out.type = .name;
        out.value = .{
            .name = try self.allocator.dupe(u8, self.buffer.items),
        };
        self.clearBuffer();
    }

    fn lexRune(self: *@This()) anyerror![]const u8 {
        const cp = try self.next(null, false);
        std.debug.assert(cp != null);
        var character = cp.?;

        switch (character) {
            '\\' => {
                character = (try self.next(null, false)).?;
                switch (character) {
                    // '0', 'a' => {
                    //     return "\\0";
                    // },
                    // 'b' => {
                    //     return .{
                    //         .characters = [_]u8{ '\b', 0, 0, 0 },
                    //         .length = 1,
                    //     };
                    // },
                    'r' => {
                        return "\r";
                    },
                    't' => {
                        return "\t";
                    },
                    'n' => {
                        return "\n";
                    },
                    // 'f' => {
                    //     return .{
                    //         .characters = [_]u8{ '\f', 0, 0, 0 },
                    //         .length = 1,
                    //     };
                    // },
                    // 'v' => {
                    //     return .{
                    //         .characters = [_]u8{ '\v', 0, 0, 0 },
                    //         .length = 1,
                    //     };
                    // },
                    '\\' => {
                        return "\\";
                    },
                    else => try self.printError("Invalid escape '\\{}'", .{character}),
                }
                util.vixUnreachable(@src());
            },
            else => {
                var encoded_result = [_]u8{0} ** 4;
                const length = try std.unicode.utf8Encode(character, &encoded_result);
                return encoded_result[0..length];
            }, // TODO
        }
    }

    fn lexString(self: *@This(), out: *Token) anyerror!void {
        var cp = try self.next(&out.location, false);
        std.debug.assert(cp != null);
        var character = cp.?;

        switch (character) {
            '<' => {
                var count: usize = 1;
                while (true) {
                    cp = try self.next(null, false);
                    if (cp == null) {
                        try self.printError("Unexpected end of file", .{});
                    }
                    character = cp.?;
                    if (character == '<') {
                        count += 1;
                    } else if (character == '>' and count == 1) {
                        break;
                    } else if (character == '>') {
                        count -= 1;
                    }
                    self.push(character, false);
                    const utf8Char = try self.lexRune();
                    for (utf8Char) |char| {
                        try self.buffer.append(char);
                    }
                }
                out.type = .string;
                out.value = .{
                    .string = try self.allocator.dupe(u8, self.buffer.items),
                };
                self.clearBuffer();
                return;
            },
            else => util.vixUnreachable(@src()),
        }
    }
};
